From 90e1911cc80af5edb425753d6b2d9ce2ca569efb Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Mon, 20 Mar 2017 16:58:49 -0400
Subject: [PATCH 96/97] net: phy: dp83867: add 4-level strap mode 1 support

The phy allows configure the device through 4 level strap. In this
configuration method, when the strap mode 1 or 2 is used, the bit
7 of CFG4 register should be cleared. This patch add support for handling
mode 1. The details of this mode are described at
http://www.ti.com/lit/ds/snls484e/snls484e.pdf

Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
Signed-off-by: Jacob Stiffler <j-stiffler@ti.com>
---
 Documentation/devicetree/bindings/net/ti,dp83867.txt |  4 ++++
 drivers/net/phy/dp83867.c                            | 17 +++++++++++++++++
 include/dt-bindings/net/ti-dp83867.h                 |  5 +++++
 3 files changed, 26 insertions(+)

diff --git a/Documentation/devicetree/bindings/net/ti,dp83867.txt b/Documentation/devicetree/bindings/net/ti,dp83867.txt
index 61d970c..b3b36ab 100644
--- a/Documentation/devicetree/bindings/net/ti,dp83867.txt
+++ b/Documentation/devicetree/bindings/net/ti,dp83867.txt
@@ -18,6 +18,9 @@ Optional property:
 	- ti,max-output-impedance - MAC Interface Impedance control to set
 				    the programmable output impedance to
 				    maximum value (70 ohms).
+	- ti,strap-mode	- 4-level strap configuration modes. When MODE1 or
+			  MODE2 is used, the CFG4 register bit 7 is cleared.
+			  currently only MODE1 is supported.
 
 Note: ti,min-output-impedance and ti,max-output-impedance are mutually
       exclusive. When both properties are present ti,max-output-impedance
@@ -33,6 +36,7 @@ Example:
 		ti,rx-internal-delay = <DP83867_RGMIIDCTL_2_25_NS>;
 		ti,tx-internal-delay = <DP83867_RGMIIDCTL_2_75_NS>;
 		ti,fifo-depth = <DP83867_PHYCR_FIFO_DEPTH_4_B_NIB>;
+		ti,strap-mode = <DP83867_4LEVEL_STRAP_MODE1>;
 	};
 
 Datasheet can be found:
diff --git a/drivers/net/phy/dp83867.c b/drivers/net/phy/dp83867.c
index fd27f71..df43cf8 100644
--- a/drivers/net/phy/dp83867.c
+++ b/drivers/net/phy/dp83867.c
@@ -31,6 +31,7 @@
 #define DP83867_CTRL		0x1f
 
 /* Extended Registers */
+#define DP83867_CFG4		0x0031
 #define DP83867_RGMIICTL	0x0032
 #define DP83867_RGMIIDCTL	0x0086
 #define DP83867_IO_MUX_CFG	0x0170
@@ -74,6 +75,7 @@ struct dp83867_private {
 	int tx_id_delay;
 	int fifo_depth;
 	int io_impedance;
+	int strap_mode;
 };
 
 static int dp83867_ack_interrupt(struct phy_device *phydev)
@@ -123,6 +125,7 @@ static int dp83867_of_init(struct phy_device *phydev)
 		return -ENODEV;
 
 	dp83867->io_impedance = -EINVAL;
+	dp83867->strap_mode = -EINVAL;
 
 	/* Optional configuration */
 	if (of_property_read_bool(of_node, "ti,max-output-impedance"))
@@ -130,6 +133,11 @@ static int dp83867_of_init(struct phy_device *phydev)
 	else if (of_property_read_bool(of_node, "ti,min-output-impedance"))
 		dp83867->io_impedance = DP83867_IO_MUX_CFG_IO_IMPEDANCE_MIN;
 
+	if (!of_property_read_u32(of_node, "ti,strap-mode",
+				  &dp83867->strap_mode))
+		if (dp83867->strap_mode != DP83867_4LEVEL_STRAP_MODE1)
+			return -EINVAL;
+
 	ret = of_property_read_u32(of_node, "ti,rx-internal-delay",
 				   &dp83867->rx_id_delay);
 	if (ret &&
@@ -174,6 +182,15 @@ static int dp83867_config_init(struct phy_device *phydev)
 		dp83867 = (struct dp83867_private *)phydev->priv;
 	}
 
+	if ((dp83867->strap_mode == DP83867_4LEVEL_STRAP_MODE1) ||
+	    (dp83867->strap_mode == DP83867_4LEVEL_STRAP_MODE2)) {
+		val = phy_read_mmd_indirect(phydev, DP83867_CFG4,
+					    DP83867_DEVADDR, phydev->addr);
+		val &= ~BIT(7);
+		phy_write_mmd_indirect(phydev, DP83867_CFG4,
+				       DP83867_DEVADDR, phydev->addr, val);
+	}
+
 	if (phy_interface_is_rgmii(phydev)) {
 		val = phy_read(phydev, MII_DP83867_PHYCTRL);
 		if (val < 0)
diff --git a/include/dt-bindings/net/ti-dp83867.h b/include/dt-bindings/net/ti-dp83867.h
index 172744a..cc8267b 100644
--- a/include/dt-bindings/net/ti-dp83867.h
+++ b/include/dt-bindings/net/ti-dp83867.h
@@ -42,4 +42,9 @@
 #define	DP83867_RGMIIDCTL_3_75_NS	0xe
 #define	DP83867_RGMIIDCTL_4_00_NS	0xf
 
+/* 4 level Strap modes */
+#define DP83867_4LEVEL_STRAP_MODE1	0x1
+#define DP83867_4LEVEL_STRAP_MODE2	0x2
+#define DP83867_4LEVEL_STRAP_MODE3	0x3
+#define DP83867_4LEVEL_STRAP_MODE4	0x4
 #endif
-- 
1.9.1

